core/src/test/java/org/apache/iceberg/TestIncrementalDataScan.java

```diff
@@ -0,0 +1,198 @@
+
+    add(transaction.newAppend(), files("A")); // 1
+    add(transaction.newAppend(), files("B"));
+    add(transaction.newAppend(), files("C"));
+    add(transaction.newAppend(), files("D"));
+    add(transaction.newAppend(), files("E")); // 5
+    transaction.commitTransaction();
+    Assert.assertEquals(Sets.newHashSet("B", "C", "D", "E"), appendsBetweenScan(1, 5));
+
+    transaction = table.newTransaction();
+    replace(transaction.newRewrite(), files("A", "B", "C"), files("F", "G")); // 6
+    transaction.commitTransaction();
+    Assert.assertEquals("Replace commits are ignored", Sets.newHashSet("B", "C", "D", "E"), appendsBetweenScan(1, 6));
+    Assert.assertEquals(Sets.newHashSet("E"), appendsBetweenScan(4, 6));
+    // 6th snapshot is a replace. No new content is added
+    Assert.assertTrue("Replace commits are ignored", appendsBetweenScan(5, 6).isEmpty());
+
+    transaction = table.newTransaction();
+    delete(transaction.newDelete(), files("D")); // 7
+    transaction.commitTransaction();
+    // 7th snapshot is a delete.
+    Assert.assertTrue("Replace and delete commits are ignored", appendsBetweenScan(5, 7).isEmpty());
+    Assert.assertTrue("Delete commits are ignored", appendsBetweenScan(6, 7).isEmpty());
+
+    transaction = table.newTransaction();
+    // 8th snapshot is an overwrite
+    overwrite(transaction.newOverwrite(), files("H"), files("E"));
+    add(transaction.newAppend(), files("I")); // 9
+    transaction.commitTransaction();
+    // snapshots 6, 7 and 8 are ignored
+    Assert.assertEquals(Sets.newHashSet("B", "C", "D", "E", "I"), appendsBetweenScan(1, 9));
+    Assert.assertEquals(Sets.newHashSet("I"), appendsBetweenScan(6, 9));
+    Assert.assertEquals(Sets.newHashSet("I"), appendsBetweenScan(7, 9));
+    Assert.assertEquals(Sets.newHashSet("I"), appendsBetweenScan(8, 9));
+  }
+
+  @Test
+  public void testRollbacks() {
+    add(table.newAppend(), files("A"));
+    add(table.newAppend(), files("B"));
+    add(table.newAppend(), files("C"));
+    // Go back to snapshot "B"
+    table.rollback().toSnapshotId(2).commit();
+    Assert.assertEquals(2, table.currentSnapshot().snapshotId());
+    Assert.assertEquals(Sets.newHashSet("B"), appendsBetweenScan(1, 2));
+
+    Transaction transaction = table.newTransaction();
+    add(transaction.newAppend(), files("D"));
+    add(transaction.newAppend(), files("E"));
+    add(transaction.newAppend(), files("F"));
+    transaction.commitTransaction();
+    // Go back to snapshot "E"
+    table.rollback().toSnapshotId(5).commit();
+    Assert.assertEquals(5, table.currentSnapshot().snapshotId());
+    Assert.assertEquals(Sets.newHashSet("B", "D", "E"), appendsBetweenScan(1, 5));
+  }
+
+  private static DataFile file(String name) {
+    return DataFiles.builder(SPEC)
+            .withPath(name + ".parquet")
+            .withFileSizeInBytes(0)
+            .withPartitionPath("data_bucket=0") // easy way to set partition data for now
+            .withRecordCount(0)
```

Comment:
I recommend making this at least 1. Otherwise, the files will be filtered out by some scan code.